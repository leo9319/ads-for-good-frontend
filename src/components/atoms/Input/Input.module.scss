@import '../../../styles/colours';
@import '../../../styles/icons-variables';

@mixin inputIcon {
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  position: absolute;
  cursor: pointer;
  user-select: none;
  width: 24px;
  height: 24px;
  right: 16px;
  top: 16px;
}

.formInput {
  position: relative;
  font-size: 18px;
  line-height: 24px;
  color: $color-neutral-600;
  user-select: none;

  input {
    font-size: 18px;
    outline: none;
    position: relative;
    height: 60px;
    border: 1px solid $color-neutral-300;
    border-radius: 9pt;
    box-sizing: border-box;
    margin-top: 0;
    padding: var(--spacing-md) var(--spacing-2xl) var(--spacing-xs)
      var(--spacing-md);
    width: 100%;
    max-width: none !important;

    &:hover,
    &:focus {
      border-color: $color-blue-500;
    }

    &:active {
      border-color: $color-blue-500;
      box-shadow: 0px 0px 0px 4px $color-button-shadow;
    }

    &::placeholder {
      color: transparent;
    }

    &::-webkit-contacts-auto-fill-button {
      visibility: hidden;
      pointer-events: none;
      position: absolute;
    }

    &:disabled {
      background-color: $color-neutral-50 !important;
    }

    &:focus {
      &::placeholder {
        color: $color-neutral-500 !important;
      }
    }
  }

  &:has(select option:checked:not([value=''])) {
    label {
      display: none;
    }
  }

  label {
    position: absolute;
    top: 16px;
    left: 18px;
    pointer-events: none;
    transform-origin: left center;
    transition: transform 250ms;
    z-index: 1;
  }

  &:has(input:focus) label,
  &:has(input:not(:placeholder-shown)) label,
  &:has(textarea:focus) label,
  &:has(textarea:not(:placeholder-shown)) label {
    transform: translateY(-100%) scale(0.75);
    top: 25px;
  }

  &:has(.errorMessage),
  &.hasError {
    input {
      border-color: $color-red-600;
    }

    label {
      color: $color-red-600;
    }
  }

  &.showPassword {
    .togglePassword {
      display: block;
      @include inputIcon();
    }

    &:has(input[type='password']) {
      .togglePassword {
        background-image: url($icon-eye-off);
      }
    }

    &:has(input[type='text']) {
      .togglePassword {
        background-image: url($icon-eye);
      }
    }
  }

  .showSuccessIcon {
    @include inputIcon();
    display: none;
  }

  &.isValid {
    input {
      border-color: $color-green-800;
    }

    .showSuccessIcon {
      display: block;
      background-image: url($icon-highlight-on);
    }
  }
}

input::-ms-reveal,
input::-ms-clear {
  display: none;
}
