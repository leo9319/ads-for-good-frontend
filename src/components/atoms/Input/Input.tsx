import React, { ReactElement, useEffect, useState } from 'react';
import {
  Control,
  FieldError,
  FieldErrors,
  RegisterOptions,
  useController,
  UseFormRegister,
  UseFormSetValue,
} from 'react-hook-form';
import TextCaption from '@components/atoms/TextCaption';
import { classNames } from '@utils/common/classNames';
import { payloadProps } from '@internal/types/common';

import styles from './Input.module.scss';

/**
 * @interface
 *
 * The InputProps type of {@link Input}.
 */
export interface InputProps
  extends React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  > {
  /**
   * Label for the input field
   * @property
   */
  label?: string;
  /**
   * Name of the input field
   * @property
   */
  name: string;
  /**
   * Value of the input field
   * @property
   */
  value?: string;
  /**
   * Function to handle the change event of the input field
   * @property
   */
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  /**
   * Placeholder for the input field
   * @property
   */
  placeholder?: string;
  /**
   * Type of the input field (text, number, email, password)
   * @property
   */
  type?: 'text' | 'number' | 'email' | 'password';
  /**
   * Required field or not
   * @defaultValue `false`
   * @property
   */
  required?: boolean;
  /**
   * ReadOnly field or not
   * @defaultValue `false`
   * @property
   */
  readOnly?: boolean;
  /**
   * Disabled field or not
   * @defaultValue `false`
   * @property
   */
  disabled?: boolean;
  /**
   * Classname for the input field
   * @property
   */
  className?: string;
  /**
   * Error message for the input field
   * @property
   */
  errorMessage?: string;
  /**
   * Register function from react-hook-form
   * @property
   */
  register?: UseFormRegister<payloadProps>;
  /**
   * Rules for the input field
   * @defaultValue `{}`
   * @property
   */
  rules?: RegisterOptions;
  /**
   * Error of the input field based on the rules to display
   * @property
   */
  error?: { [key: string]: FieldError } | FieldErrors;
  /**
   * Status of the HTTP request
   *  @property {number} status - The HTTP status code or application-defined status (e.g., 404, 500).
   */
  status?: number;
  /**
   * Control from react-hook-form (optional)
   * @property
   */
  control?: Control<payloadProps>;
  /**
   * showSuccessIcon field or not
   * @defaultValue `false`
   * @property
   */
  showSuccessIcon?: boolean;
  /**
   * Function to update the input keys in the useForm hook
   * @property
   */
  setValue?: UseFormSetValue<payloadProps>;
}

/**
 * Input component to display input fields
 *
 * @group Atoms
 * @category Component
 *
 * @param {InputProps} props - The Input props
 * @returns {ReactElement} - Input Component
 */
export const Input = ({
  label,
  name,
  register,
  onChange,
  rules,
  error = {},
  status,
  type = 'text',
  placeholder = '',
  className = '',
  control,
  showSuccessIcon = false,
  setValue,
  ...props
}: InputProps): ReactElement => {
  const [inputType, setInputType] = useState(type);
  const [mode, setMode] = useState<'success' | 'error'>('error');
  const {
    fieldState: { invalid = true, isTouched = false },
  } = control ? useController({ name, control, rules }) : { fieldState: {} };

  useEffect(() => {
    setMode(status && status >= 200 && status < 300 ? 'success' : 'error');
  }, [error]);

  const handleTogglePassword = () => {
    setInputType(inputType === 'text' ? 'password' : 'text');
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (setValue) {
      setValue(name, e.target.value, { shouldDirty: true, shouldTouch: true });
    }
    onChange?.(e);
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && setValue) {
      const inputElement = e.target as HTMLInputElement;
      setValue(name, inputElement.value, { shouldValidate: true });
    }
  };

  const checkPassword = type === 'password' ? styles.showPassword : '';
  const checkError = (error[name] as FieldError)?.message
    ? styles.hasError
    : '';

  const checkSuccess = isTouched && !invalid ? styles.isValid : '';
  const classes = classNames(
    styles.formInput,
    className,
    checkPassword,
    checkError,
    checkSuccess
  );

  return (
    <div className={classes}>
      {label && <label htmlFor={name}>{label}</label>}
      <input
        id={name}
        type={inputType}
        placeholder={placeholder}
        {...(name && register
          ? register(name, {
              ...rules,
              onChange: handleChange,
            })
          : {})}
        {...props}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
      />
      {(showSuccessIcon || type === 'password') && (
        <div
          className={
            isTouched && !invalid
              ? styles.showSuccessIcon
              : type === 'password'
                ? styles.togglePassword
                : ''
          }
          onClick={type === 'password' ? handleTogglePassword : () => {}}
        ></div>
      )}
      {error[name] && (
        <TextCaption
          messages={(error[name] as FieldError)?.message}
          mode={mode}
        />
      )}
    </div>
  );
};

export default Input;
