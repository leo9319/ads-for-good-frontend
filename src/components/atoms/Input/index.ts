export type { InputProps } from './Input';
export { default } from './Input';
export { Input } from './Input';
