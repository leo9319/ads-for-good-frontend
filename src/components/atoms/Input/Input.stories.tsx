import React from 'react';
import { FieldValues, useForm, UseFormRegister } from 'react-hook-form';
import type { Meta, StoryObj } from '@storybook/react';
import { Theme } from '@radix-ui/themes';
import { Input } from './Input';
import { payloadProps } from '@internal/types/common';

const meta = {
  title: 'Library/Atoms/Input',
  component: Input,
  decorators: [
    Story => (
      <Theme>
        <Story />
      </Theme>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: { control: 'text' },
    name: { control: 'text' },
    type: {
      control: 'select',
      options: ['text', 'number', 'email', 'password'],
    },
    placeholder: { control: 'text' },
    readOnly: { control: 'boolean' },
    disabled: { control: 'boolean' },
    required: { control: 'boolean' },
    className: { control: 'text' },
    errorMessage: { control: 'text' },
    showSuccessIcon: { control: 'boolean' },
  },
  args: {
    label: 'Input Label',
    name: 'example',
    type: 'text',
    placeholder: 'Enter text here',
    readOnly: false,
    disabled: false,
  },
} satisfies Meta<typeof Input>;

export default meta;
type Story = StoryObj<typeof meta>;

export const DefaultInput: Story = {
  args: {
    label: 'Default Input',
    name: 'example',
  },
};

export const EmailInput: Story = {
  args: {
    label: 'Email Input',
    type: 'email',
    placeholder: 'Enter email address',
    name: 'email',
    rules: {
      required: 'Email is required',
      pattern: {
        value: /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/,
        message: 'Enter a valid email address',
      },
    },
  },
};

export const PasswordInput: Story = {
  args: {
    label: 'Password Input',
    type: 'password',
    placeholder: 'Enter Password',
    name: 'password',
    rules: {
      required: 'Email is required',
    },
  },
};

export const EmailValidationInput: Story = {
  render: function EmailValidationComponent() {
    const {
      register,
      handleSubmit,
      setValue,
      control,
      formState: { errors },
    } = useForm<payloadProps>({
      mode: 'onBlur', // Validate on blur
    });

    const onSubmit = (data: payloadProps) => {
      console.log('Form Data:', data);
    };

    return (
      <form onSubmit={handleSubmit(onSubmit)} noValidate>
        <Input
          label="Email Input"
          name="email"
          type="text"
          placeholder="Enter your email"
          register={register as unknown as UseFormRegister<FieldValues>}
          control={control}
          showSuccessIcon={true}
          rules={{
            required: 'Email is required',
            pattern: {
              value: /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/,
              message: 'Enter a valid email address',
            },
          }}
          error={errors}
          setValue={setValue}
        />
      </form>
    );
  },
};

export const PasswordValidationInput: Story = {
  render: function PasswordValidationComponent() {
    const {
      register,
      handleSubmit,
      setValue,
      control,
      formState: { errors },
    } = useForm<payloadProps>({
      mode: 'onBlur', // Validate on blur
    });

    const onSubmit = (data: payloadProps) => {
      console.log('Form Data:', data);
    };

    return (
      <form onSubmit={handleSubmit(onSubmit)} noValidate>
        <Input
          label="Password Input"
          name="password"
          type="password"
          placeholder="Enter your password"
          register={register as unknown as UseFormRegister<FieldValues>}
          control={control}
          showSuccessIcon={true}
          rules={{
            required: 'Password is required',
            pattern: {
              value: /^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9\W_]).{8,}$/,
              message: 'Enter a valid password',
            },
          }}
          error={errors}
          setValue={setValue}
        />
      </form>
    );
  },
};
