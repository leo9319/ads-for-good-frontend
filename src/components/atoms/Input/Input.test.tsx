import React from 'react';
import Input from './Input';
import { render } from '@testing-library/react';

describe('Input', () => {
  it('should render successfully', () => {
    const { baseElement } = render(<Input name="test-1" />);
    expect(baseElement).toMatchSnapshot();
  });
});
