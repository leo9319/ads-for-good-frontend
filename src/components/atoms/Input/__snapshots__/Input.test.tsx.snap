// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render successfully 1`] = `
<body>
  <div>
    <div
      class="formInput"
    >
      <input
        id="test-1"
        placeholder=""
        type="text"
      />
    </div>
  </div>
</body>
`;
